typing/typecore.ml

```diff
@@ -4198,42 +4372,64 @@ and type_expect_
         raise (Typetexp.Error (loc, env,
           Unsupported_extension Polymorphic_parameters));
       type_function ?in_function loc sexp.pexp_attributes env
-                    expected_mode ty_expected_explained l ~has_local
+                    expected_mode ty_expected_explained l ~mode_annots
                     ~has_poly [Ast_helper.Exp.case spat sbody]
   | Pexp_function caselist ->
+      let mode_annots = mode_annots_none in
       type_function ?in_function
         loc sexp.pexp_attributes env expected_mode
-        ty_expected_explained Nolabel ~has_local:false ~has_poly:false caselist
+        ty_expected_explained Nolabel ~mode_annots ~has_poly:false caselist
   | Pexp_apply
-      ({ pexp_desc = Pexp_extension({txt = ("ocaml.local" | "local" | "extension.local" as txt)}, PStr []) },
+      ({ pexp_desc = Pexp_extension({
+            txt = ("ocaml.unique" | "unique" | "extension.unique" as txt)}, PStr []) },
        [Nolabel, sbody]) ->
-      if txt = "extension.local" && not (Language_extension.is_enabled Local) then
-        raise (Typetexp.Error (loc, Env.empty, Unsupported_extension Local));
-
-      let mode = if mode_cross env ty_expected then
-        (* when mode crosses, we check the inner expr with the most relaxed mode *)
-        {expected_mode with mode = Value_mode.local; exact = false}
-        (* moreover, because mode crosses, expected_mode is completely useless *)
-      else begin
-        (* if mode does not cross, expected.mode must be local *)
-        submode ~loc ~env ~reason:Other Value_mode.local expected_mode;
-        (* and we require the inner expr to be exact local *)
-        {expected_mode with mode = Value_mode.local; exact = true}
-      end
+      if txt = "extension.unique" && not (Language_extension.is_enabled Unique) then
+        raise (Typetexp.Error (loc, Env.empty, Unsupported_extension Unique));
+      let expected_mode = mode_unique expected_mode in
```

Comment:
I now want to add back the line `submode unique expected_mode` before this line. It is true that this would be a no-op, but `typecore.ml` doesn't need to know that `unique` the least mode in the axis? I think currently `typecore.ml` doesn't assume `global` is min and `local` is max.

Reply:
It's not about it being the least mode it's about the meaning of `unique_ e`. It is a request to ensure that the `e` is at least `unique_`. Consider what `exclusive_ e` should mean when the result of `e` is `unique` and is then used uniquely. I would say that should not be an error: `exclusive_ e` is a request to ensure that the result of `e` is at least `exclusive` not a request that the result of the expression be at most `exclusive.

Reply:
> It is a request to ensure that the e is at least unique_

I guess you mean "e is stricter than `unique`"? But in any case, the original `expected_mode` shouldn't just be ignored (for soundness), except here because `unique` is the lowest mode?

Reply:
Right, `mode_unique expected_mode` should return the meet of `expected_mode` and `unique`, which is always `unique` because `unique` is the lowest mode.

Reply:
But with the same logic, `once_ e` would be useless? Because `once` is already the most relaxed mode in that axis. Or does it mean "relax `e` to `once` and error if later used as `many`"?

Reply:
`once_` is interpreted in the other direction, as a requirement on the `expected_mode`. My point is that I think we should choose to interpret different modes differently when used as expressions. I think the distinction is where the mode sits in its axis relative to the legacy mode. If a mode is stricter that legacy then it should be interpreted as a requirement on the expression, if a mode is less strict than legacy then it should be interpreted as a requirement on the context.